File: mac/Sources/Transcriber/Library.swift
import Foundation

/// A user-created grouping of transcripts, like a ChatGPT/Claude project.
struct Project: Identifiable, Codable, Hashable {
    var id: UUID = UUID()
    var name: String
    var notes: String = ""
    var created: Date = Date()
}

/// One transcript in the managed library. The transcript `.md` and (optionally) a copy of the
/// source audio live in `Library/<id>/`; the fields below are denormalised from the transcript's
/// frontmatter so the list can render without opening every file.
struct LibraryItem: Identifiable, Codable, Hashable {
    var id: UUID = UUID()
    var projectID: UUID?          // nil == the built-in "Unsorted" bucket
    var title: String
    var transcriptFile: String    // filename inside Library/<id>/
    var audioFile: String?        // filename inside Library/<id>/, if audio was copied in
    var created: Date = Date()
    var modified: Date = Date()

    // Denormalised for display.
    var durationSeconds: Double?
    var language: String?
    var speakers: [String] = []
    var hasSummary: Bool = false
    var recordingWarning: String?
    var sourceName: String?
}

/// The on-disk index. Bumping `version` lets future migrations detect an old layout.
struct LibraryIndex: Codable {
    var version: Int = 1
    var projects: [Project] = []
    var items: [LibraryItem] = []
}

/// Owns the managed library: the `Library/` folder, its `library.json` index, and every
/// mutation. The library is the source of truth for transcripts — files are copied *in* on
/// ingest and only ever leave through `export`. Deliberately holds no reference to `AppState`
/// so the same store can be driven from tests and (later) the MCP process.
final class LibraryStore: ObservableObject {
    @Published private(set) var index = LibraryIndex()

    let root: URL

    /// - Parameter root: the `Library` directory. Tests pass a temp dir for isolation.
    init(root: URL = AppPaths.support.appendingPathComponent("Library", isDirectory: true)) {
        self.root = root
        try? FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
        load()
    }

    var indexURL: URL { root.appendingPathComponent("library.json") }
    private func itemFolder(_ id: UUID) -> URL { root.appendingPathComponent(id.uuidString, isDirectory: true) }

    // MARK: - Convenience reads

    var projects: [Project] { index.projects.sorted { $0.created < $1.created } }
    var items: [LibraryItem] { index.items.sorted { $0.modified > $1.modified } }

    func items(in projectID: UUID?) -> [LibraryItem] {
        items.filter { $0.projectID == projectID }
    }

    func transcriptURL(for item: LibraryItem) -> URL {
        itemFolder(item.id).appendingPathComponent(item.transcriptFile)
    }

    func audioURL(for item: LibraryItem) -> URL? {
        item.audioFile.map { itemFolder(item.id).appendingPathComponent($0) }
    }

    // MARK: - Projects

    @discardableResult
    func createProject(name: String, notes: String = "") -> Project {
        let project = Project(name: name.trimmingCharacters(in: .whitespaces), notes: notes)
        index.projects.append(project)
        save()
        return project
    }

    func renameProject(_ id: UUID, to name: String) {
        guard let i = index.projects.firstIndex(where: { $0.id == id }) else { return }
        index.projects[i].name = name.trimmingCharacters(in: .whitespaces)
        save()
    }

    func setNotes(_ id: UUID, notes: String) {
        guard let i = index.projects.firstIndex(where: { $0.id == id }) else { return }
        index.projects[i].notes = notes
        save()
    }

    /// Deleting a project keeps its transcripts — they fall back to Unsorted.
    func deleteProject(_ id: UUID) {
        for i in index.items.indices where index.items[i].projectID == id {
            index.items[i].projectID = nil
        }
        index.projects.removeAll { $0.id == id }
        save()
    }

    // MARK: - Items

    /// Copies a transcript (and optionally its audio) into the library and records it.
    /// The originals are never modified.
    @discardableResult
    func ingest(
        transcriptURL: URL,
        audioURL: URL? = nil,
        projectID: UUID? = nil,
        title: String? = nil,
        copyAudio: Bool = true
    ) throws -> LibraryItem {
        let id = UUID()
        let folder = itemFolder(id)
        try FileManager.default.createDirectory(at: folder, withIntermediateDirectories: true)

        let displayTitle = title ?? transcriptURL.deletingPathExtension().lastPathComponent
        let mdName = Self.sanitize(displayTitle) + ".md"
        let mdDestination = folder.appendingPathComponent(mdName)
        try FileManager.default.copyItem(at: transcriptURL, to: mdDestination)

        var audioName: String?
        if copyAudio, let audioURL, FileManager.default.fileExists(atPath: audioURL.path) {
            let name = Self.sanitize(displayTitle) + "." + audioURL.pathExtension
            let destination = folder.appendingPathComponent(name)
            try? FileManager.default.copyItem(at: audioURL, to: destination)
            if FileManager.default.fileExists(atPath: destination.path) { audioName = name }
        }

        var item = LibraryItem(
            id: id,                       // must match the folder created above
            projectID: projectID,
            title: displayTitle,
            transcriptFile: mdName,
            audioFile: audioName,
            sourceName: audioURL?.lastPathComponent
        )
        applyMetadata(to: &item, from: mdDestination)

        index.items.append(item)
        save()
        return item
    }

    /// Refreshes the denormalised fields from the transcript on disk (call after editing the .md).
    func refreshMetadata(_ id: UUID) {
        guard let i = index.items.firstIndex(where: { $0.id == id }) else { return }
        applyMetadata(to: &index.items[i], from: transcriptURL(for: index.items[i]))
        index.items[i].modified = Date()
        save()
    }

    func move(_ id: UUID, to projectID: UUID?) {
        guard let i = index.items.firstIndex(where: { $0.id == id }) else { return }
        index.items[i].projectID = projectID
        index.items[i].modified = Date()
        save()
    }

    /// Renames the item: moves the transcript `.md` (and its sibling audio) on disk, rewrites the
    /// `# heading` inside the transcript, and updates the record. Renaming the *file* is what keeps
    /// search and the MCP server consistent, since both title a transcript by its filename.
    @discardableResult
    func rename(_ id: UUID, to newTitle: String) -> Bool {
        guard let i = index.items.firstIndex(where: { $0.id == id }) else { return false }
        let clean = newTitle.trimmingCharacters(in: .whitespaces)
        guard !clean.isEmpty else { return false }

        let folder = itemFolder(id)
        let currentMD = index.items[i].transcriptFile
        let base = uniqueBase(Self.sanitize(clean), in: folder, keeping: currentMD)

        // Move the transcript. If this fails we abort without touching the record.
        let oldMD = folder.appendingPathComponent(currentMD)
        let newMDName = base + ".md"
        let newMD = folder.appendingPathComponent(newMDName)
        if newMD != oldMD {
            do {
                try FileManager.default.moveItem(at: oldMD, to: newMD)
            } catch {
                Log.recording.error("rename: could not move transcript: \(error.localizedDescription, privacy: .public)")
                return false
            }
            index.items[i].transcriptFile = newMDName
        }

        // Move the sibling audio to match, keeping its extension. Best-effort.
        if let audio = index.items[i].audioFile {
            let oldAudio = folder.appendingPathComponent(audio)
            let ext = (audio as NSString).pathExtension
            let newAudioName = base + (ext.isEmpty ? "" : "." + ext)
            let newAudio = folder.appendingPathComponent(newAudioName)
            if newAudio != oldAudio, FileManager.default.fileExists(atPath: oldAudio.path) {
                try? FileManager.default.moveItem(at: oldAudio, to: newAudio)
                if FileManager.default.fileExists(atPath: newAudio.path) {
                    index.items[i].audioFile = newAudioName
                }
            }
        }

        // Keep the in-file H1 in sync so the transcript reads right when opened directly.
        rewriteHeading(at: newMD, to: clean)

        index.items[i].title = clean
        index.items[i].modified = Date()
        save()
        return true
    }

    /// Picks a filename base that doesn't collide with another file in `folder`, ignoring the
    /// item's own current transcript. Appends " 2", " 3", … on collision.
    private func uniqueBase(_ base: String, in folder: URL, keeping currentMD: String) -> String {
        let fm = FileManager.default
        var candidate = base
        var n = 2
        while candidate + ".md" != currentMD,
              fm.fileExists(atPath: folder.appendingPathComponent(candidate + ".md").path) {
            candidate = "\(base) \(n)"
            n += 1
        }
        return candidate
    }

    /// Replaces the first Markdown H1 (`# …`) in the transcript with `title`. No-op if absent.
    private func rewriteHeading(at url: URL, to title: String) {
        guard var content = try? String(contentsOf: url, encoding: .utf8) else { return }
        var lines = content.components(separatedBy: "\n")
        guard let idx = lines.firstIndex(where: { $0.hasPrefix("# ") }) else { return }
        lines[idx] = "# " + title
        content = lines.joined(separator: "\n")
        try? content.write(to: url, atomically: true, encoding: .utf8)
    }

    /// Permanently removes the library's copy (transcript + audio). Never touches anything the
    /// user has elsewhere on disk.
    func deleteItem(_ id: UUID) {
        try? FileManager.default.removeItem(at: itemFolder(id))
        index.items.removeAll { $0.id == id }
        save()
    }

    enum ExportKind { case transcript, audio }

    /// Copies the stored transcript or audio out to a user-chosen location and name.
    /// Returns the written URL. Duplicates are expected and fine.
    @discardableResult
    func export(_ id: UUID, kind: ExportKind, to destination: URL) throws -> URL {
        guard let item = index.items.first(where: { $0.id == id }) else {
            throw LibraryError.notFound
        }
        let source: URL?
        switch kind {
        case .transcript: source = transcriptURL(for: item)
        case .audio: source = audioURL(for: item)
        }
        guard let source, FileManager.default.fileExists(atPath: source.path) else {
            throw LibraryError.missingFile
        }
        if FileManager.default.fileExists(atPath: destination.path) {
            try FileManager.default.removeItem(at: destination)
        }
        try FileManager.default.copyItem(at: source, to: destination)
        return destination
    }

    // MARK: - Migration

    private var migrationMarker: URL { root.appendingPathComponent(".migrated") }

    /// One-time import of transcripts produced before the library existed. Copies each `.md`
    /// (and its sibling audio if still present) into an "Unsorted" item. Idempotent: a marker
    /// file ensures it runs once, and it never moves or deletes the originals.
    func migrateIfNeeded(transcriptPaths: [URL], audioResolver: (URL) -> URL? = { _ in nil }) {
        guard !FileManager.default.fileExists(atPath: migrationMarker.path) else { return }

        for md in transcriptPaths where FileManager.default.fileExists(atPath: md.path) {
            let audio = audioResolver(md)
            _ = try? ingest(transcriptURL: md, audioURL: audio, projectID: nil)
        }
        FileManager.default.createFile(atPath: migrationMarker.path, contents: nil)
        save()
    }

    // MARK: - Persistence

    private func load() {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601   // must match save()'s encoder
        guard let data = try? Data(contentsOf: indexURL),
              let decoded = try? decoder.decode(LibraryIndex.self, from: data) else { return }
        index = decoded
    }

    private func save() {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        encoder.dateEncodingStrategy = .iso8601
        guard let data = try? encoder.encode(index) else { return }
        try? data.write(to: indexURL, options: .atomic)
    }

    // MARK: - Helpers

    private func applyMetadata(to item: inout LibraryItem, from md: URL) {
        guard let entry = TranscriptIndex.load(md) else { return }
        item.language = entry.languages.joined(separator: ", ").ifEmptyNil
        item.speakers = entry.speakers
        item.hasSummary = entry.hasSummary
        item.durationSeconds = Self.seconds(from: entry.duration)
        item.recordingWarning = entry.recordingWarningText
        if item.sourceName == nil { item.sourceName = entry.sourceName }
        item.modified = entry.modified ?? item.modified
    }

    /// "44:42" / "1:07:06" -> seconds.
    static func seconds(from label: String?) -> Double? {
        guard let label else { return nil }
        let parts = label.split(separator: ":").map { Double($0) ?? 0 }
        switch parts.count {
        case 3: return parts[0] * 3600 + parts[1] * 60 + parts[2]
        case 2: return parts[0] * 60 + parts[1]
        case 1: return parts[0]
        default: return nil
        }
    }

    /// Filesystem-safe filename component (mirrors FileRenamer's rules).
    static func sanitize(_ name: String) -> String {
        let cleaned = name.components(separatedBy: CharacterSet(charactersIn: "/\\:*?\"<>|"))
            .joined(separator: " ")
            .trimmingCharacters(in: .whitespaces)
        return cleaned.isEmpty ? "transcript" : cleaned
    }

    enum LibraryError: Error { case notFound, missingFile }
}

private extension String {
    var ifEmptyNil: String? { isEmpty ? nil : self }
}

private extension TranscriptEntry {
    /// The library stores `recording_warning:` as plain metadata; TranscriptEntry doesn't parse
    /// it, so pull it straight from the frontmatter when present.
    var recordingWarningText: String? {
        guard let data = try? String(contentsOf: url, encoding: .utf8) else { return nil }
        return TranscriptIndex.frontmatter(data)["recording_warning"]
    }
}

File: windows/src/Transcriber.Core/Library.cs
using System.Text.Json;

namespace Transcriber.Core;

/// <summary>A user-created grouping of transcripts, like a ChatGPT/Claude project.</summary>
public sealed class Project
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string Name { get; set; } = string.Empty;
    public string Notes { get; set; } = string.Empty;
    public DateTimeOffset Created { get; set; } = DateTimeOffset.UtcNow;
}

/// <summary>
/// One transcript in the managed library. The transcript `.md` and (optionally) a copy of the
/// source audio live in <c>Library/&lt;id&gt;/</c>; the remaining fields are denormalised from the
/// transcript's frontmatter so a list can render without opening every file.
/// </summary>
public sealed class LibraryItem
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid? ProjectId { get; set; }          // null == the built-in "Unsorted" bucket
    public string Title { get; set; } = string.Empty;
    public string TranscriptFile { get; set; } = string.Empty;
    public string? AudioFile { get; set; }
    public DateTimeOffset Created { get; set; } = DateTimeOffset.UtcNow;
    public DateTimeOffset Modified { get; set; } = DateTimeOffset.UtcNow;

    public double? DurationSeconds { get; set; }
    public string? Language { get; set; }
    public List<string> Speakers { get; set; } = [];
    public bool HasSummary { get; set; }
    public string? RecordingWarning { get; set; }
    public string? SourceName { get; set; }
}

public sealed class LibraryIndex
{
    public int Version { get; set; } = 1;
    public List<Project> Projects { get; set; } = [];
    public List<LibraryItem> Items { get; set; } = [];
}

public enum ExportKind { Transcript, Audio }

public sealed class LibraryException(string message) : Exception(message);

/// <summary>
/// The managed library: the <c>Library/</c> folder, its <c>library.json</c> index, and every
/// mutation. Mirrors the macOS <c>LibraryStore.swift</c> (see FORMAT.md) so both products behave
/// identically. Files are copied *in* on ingest and only ever leave through <see cref="Export"/>.
/// </summary>
public sealed class LibraryStore
{
    private static readonly JsonSerializerOptions JsonOptions = new() { WriteIndented = true };

    public string Root { get; }
    private LibraryIndex _index = new();

    public LibraryStore(string root)
    {
        Root = root;
        Directory.CreateDirectory(root);
        Load();
    }

    public string IndexPath => Path.Combine(Root, "library.json");
    private string ItemFolder(Guid id) => Path.Combine(Root, id.ToString("D"));

    // --- reads ---------------------------------------------------------------------------

    public IReadOnlyList<Project> Projects => _index.Projects.OrderBy(p => p.Created).ToList();
    public IReadOnlyList<LibraryItem> Items => _index.Items.OrderByDescending(i => i.Modified).ToList();
    public IReadOnlyList<LibraryItem> ItemsIn(Guid? projectId) =>
        Items.Where(i => i.ProjectId == projectId).ToList();

    public string TranscriptPath(LibraryItem item) => Path.Combine(ItemFolder(item.Id), item.TranscriptFile);
    public string? AudioPath(LibraryItem item) =>
        item.AudioFile is null ? null : Path.Combine(ItemFolder(item.Id), item.AudioFile);

    // --- projects ------------------------------------------------------------------------

    public Project CreateProject(string name, string notes = "")
    {
        var project = new Project { Name = name.Trim(), Notes = notes };
        _index.Projects.Add(project);
        Save();
        return project;
    }

    public void RenameProject(Guid id, string name)
    {
        var project = _index.Projects.FirstOrDefault(p => p.Id == id);
        if (project is null) return;
        project.Name = name.Trim();
        Save();
    }

    public void SetNotes(Guid id, string notes)
    {
        var project = _index.Projects.FirstOrDefault(p => p.Id == id);
        if (project is null) return;
        project.Notes = notes;
        Save();
    }

    /// <summary>Deleting a project keeps its transcripts — they fall back to Unsorted.</summary>
    public void DeleteProject(Guid id)
    {
        foreach (var item in _index.Items.Where(i => i.ProjectId == id))
        {
            item.ProjectId = null;
        }
        _index.Projects.RemoveAll(p => p.Id == id);
        Save();
    }

    // --- items ---------------------------------------------------------------------------

    /// <summary>
    /// Copies a transcript (and optionally its audio) into the library and records it. The
    /// originals are never modified.
    /// </summary>
    public LibraryItem Ingest(
        string transcriptPath,
        string? audioPath = null,
        Guid? projectId = null,
        string? title = null,
        bool copyAudio = true)
    {
        var id = Guid.NewGuid();
        var folder = ItemFolder(id);
        Directory.CreateDirectory(folder);

        var displayTitle = title ?? Path.GetFileNameWithoutExtension(transcriptPath);
        var mdName = Sanitize(displayTitle) + ".md";
        var mdDestination = Path.Combine(folder, mdName);
        File.Copy(transcriptPath, mdDestination, overwrite: true);

        string? audioName = null;
        if (copyAudio && audioPath is not null && File.Exists(audioPath))
        {
            var name = Sanitize(displayTitle) + Path.GetExtension(audioPath);
            var destination = Path.Combine(folder, name);
            try
            {
                File.Copy(audioPath, destination, overwrite: true);
                if (File.Exists(destination)) audioName = name;
            }
            catch { /* audio copy is best-effort */ }
        }

        var item = new LibraryItem
        {
            Id = id,
            ProjectId = projectId,
            Title = displayTitle,
            TranscriptFile = mdName,
            AudioFile = audioName,
            SourceName = audioPath is null ? null : Path.GetFileName(audioPath),
        };
        ApplyMetadata(item, mdDestination);

        _index.Items.Add(item);
        Save();
        return item;
    }

    /// <summary>Refreshes denormalised fields from the transcript on disk (after editing it).</summary>
    public void RefreshMetadata(Guid id)
    {
        var item = _index.Items.FirstOrDefault(i => i.Id == id);
        if (item is null) return;
        ApplyMetadata(item, TranscriptPath(item));
        item.Modified = DateTimeOffset.UtcNow;
        Save();
    }

    public void Move(Guid id, Guid? projectId)
    {
        var item = _index.Items.FirstOrDefault(i => i.Id == id);
        if (item is null) return;
        item.ProjectId = projectId;
        item.Modified = DateTimeOffset.UtcNow;
        Save();
    }

    public void SetTitle(Guid id, string title)
    {
        var item = _index.Items.FirstOrDefault(i => i.Id == id);
        if (item is null) return;
        item.Title = title.Trim();
        item.Modified = DateTimeOffset.UtcNow;
        Save();
    }

    /// <summary>
    /// Permanently removes the library's copy (transcript + audio). Never touches anything the
    /// user has elsewhere on disk.
    /// </summary>
    public void DeleteItem(Guid id)
    {
        try { Directory.Delete(ItemFolder(id), recursive: true); } catch { /* already gone */ }
        _index.Items.RemoveAll(i => i.Id == id);
        Save();
    }

    /// <summary>Copies the stored transcript or audio out to a chosen location and name.</summary>
    public string Export(Guid id, ExportKind kind, string destination)
    {
        var item = _index.Items.FirstOrDefault(i => i.Id == id)
            ?? throw new LibraryException("transcript not found");
        var source = kind switch
        {
            ExportKind.Transcript => TranscriptPath(item),
            ExportKind.Audio => AudioPath(item),
            _ => null,
        };
        if (source is null || !File.Exists(source))
        {
            throw new LibraryException("the file to export is missing");
        }
        File.Copy(source, destination, overwrite: true);
        return destination;
    }

    // --- migration -----------------------------------------------------------------------

    private string MigrationMarker => Path.Combine(Root, ".migrated");

    /// <summary>
    /// One-time import of transcripts produced before the library existed. Copies each `.md`
    /// (and its resolved audio) into an "Unsorted" item. Idempotent via a marker file; never
    /// moves or deletes the originals.
    /// </summary>
    public void MigrateIfNeeded(IEnumerable<string> transcriptPaths, Func<string, string?>? audioResolver = null)
    {
        if (File.Exists(MigrationMarker)) return;

        foreach (var md in transcriptPaths.Where(File.Exists))
        {
            var audio = audioResolver?.Invoke(md);
            try { Ingest(md, audio); } catch { /* skip unreadable transcripts */ }
        }
        File.WriteAllText(MigrationMarker, string.Empty);
        Save();
    }

    // --- helpers -------------------------------------------------------------------------

    private void ApplyMetadata(LibraryItem item, string mdPath)
    {
        string content;
        try { content = File.ReadAllText(mdPath); } catch { return; }

        var fields = TranscriptFrontmatter.Parse(content);
        if (fields.TryGetValue("language", out var language) && language.Length > 0)
        {
            item.Language = language;
        }
        item.Speakers = TranscriptFrontmatter.ParseList(fields.GetValueOrDefault("speakers"));
        item.DurationSeconds = Seconds(fields.GetValueOrDefault("duration"));
        item.RecordingWarning = fields.GetValueOrDefault("recording_warning");
        item.HasSummary = TranscriptFrontmatter.HasSummary(content);
        item.SourceName ??= fields.GetValueOrDefault("source");
    }

    /// <summary>"44:42" / "1:07:06" → seconds.</summary>
    public static double? Seconds(string? label)
    {
        if (string.IsNullOrWhiteSpace(label)) return null;
        var parts = label.Split(':');
        double total = 0;
        foreach (var part in parts)
        {
            if (!double.TryParse(part, System.Globalization.NumberStyles.Float,
                    System.Globalization.CultureInfo.InvariantCulture, out var value))
            {
                return null;
            }
            total = total * 60 + value;
        }
        return total;
    }

    /// <summary>Filesystem-safe filename component (mirrors the Swift FileRenamer rules).</summary>
    public static string Sanitize(string name)
    {
        // Replace illegal characters with spaces, then collapse whitespace runs so we don't
        // leave awkward double spaces where a "x: y" became "x  y".
        var withoutIllegal = string.Join(" ", name.Split('/', '\\', ':', '*', '?', '"', '<', '>', '|'));
        var collapsed = string.Join(" ", withoutIllegal.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries));
        return collapsed.Length == 0 ? "transcript" : collapsed;
    }

    private void Load()
    {
        if (!File.Exists(IndexPath)) return;
        try
        {
            var decoded = JsonSerializer.Deserialize<LibraryIndex>(File.ReadAllText(IndexPath), JsonOptions);
            if (decoded is not null) _index = decoded;
        }
        catch { /* corrupt index → start fresh rather than crash */ }
    }

    private void Save()
    {
        var json = JsonSerializer.Serialize(_index, JsonOptions);
        File.WriteAllText(IndexPath, json);
    }
}

/// <summary>
/// Reads the YAML frontmatter this app writes (a fixed, simple subset — not general YAML).
/// Mirrors <c>TranscriptIndex.frontmatter</c> on macOS so the two products read each other's files.
/// </summary>
public static class TranscriptFrontmatter
{
    public const string SummaryStart = "<!-- SUMMARY:START -->";

    public static Dictionary<string, string> Parse(string content)
    {
        var fields = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        if (!content.StartsWith("---")) return fields;

        // Everything between the opening "---" line and the next "---" line.
        var lines = content.Replace("\r\n", "\n").Split('\n');
        for (var i = 1; i < lines.Length; i++)
        {
            var line = lines[i];
            if (line.Trim() == "---") break;
            var colon = line.IndexOf(':');
            if (colon <= 0) continue;
            var key = line[..colon].Trim();
            var value = line[(colon + 1)..].Trim();
            // Strip surrounding quotes from scalar string values.
            if (value.Length >= 2 && value.StartsWith('"') && value.EndsWith('"'))
            {
                value = value[1..^1];
            }
            fields[key] = value;
        }
        return fields;
    }

    /// <summary>Parses a `["a", "b"]` frontmatter list, or a comma-separated fallback.</summary>
    public static List<string> ParseList(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw)) return [];
        var trimmed = raw.Trim();
        if (trimmed.StartsWith('[') && trimmed.EndsWith(']'))
        {
            trimmed = trimmed[1..^1];
        }
        return trimmed
            .Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            .Select(part => part.Trim().Trim('"').Trim())
            .Where(part => part.Length > 0)
            .ToList();
    }

    public static bool HasSummary(string content) => content.Contains(SummaryStart);
}
